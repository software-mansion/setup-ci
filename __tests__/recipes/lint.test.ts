import { readFileSync, existsSync } from 'fs'
import { join } from 'path'
import {
  cli,
  getPackageJsonWithoutVersions,
  PRESET_FLAG,
  removeTestProject,
  setupTestProject,
  TEST_PROJECTS,
} from '../utils'

const FLAG = 'lint'
const PRETTIER_FLAG = 'prettier'

describe('lint recipe', () => {
  afterEach(removeTestProject)

  const PROJECTS = [
    'rn-setup-ci-yarn-flat',
    'rn-setup-ci-npm-flat',
    'rn-setup-ci-yarn-monorepo',
    'rn-setup-ci-npm-monorepo',
    'rn-setup-ci-create-expo-stack',
    'rn-setup-ci-create-expo-stack-bun',
    'rn-setup-ci-create-expo-stack-pnpm',
  ]

  for (const projectName of PROJECTS) {
    test(`${projectName} lint`, async () => {
      setupTestProject(projectName)

      const {
        packageManager,
        repoRoot,
        appRoot,
        workflowNamePrefix,
        existingConfig,
      } = TEST_PROJECTS[projectName]

      const output = await cli([PRESET_FLAG, `--${FLAG}`], {
        cwd: appRoot,
      })

      for (const message of [
        `Detected ${packageManager} as your package manager.`,
        'Generating ESLint workflow',
        'Created ESLint workflow.',
      ]) {
        expect(output).toContain(message)
      }

      const packageJson = await getPackageJsonWithoutVersions(
        join(appRoot, 'package.json')
      )
      expect(packageJson).toMatchSnapshot()

      const filesToCheck = [
        join(repoRoot, '.github', 'workflows', `${workflowNamePrefix}lint.yml`),
      ]

      if (!existingConfig?.eslint) {
        filesToCheck.push(join(appRoot, '.eslintrc.json'))
      } else if (existingConfig.eslint !== '.eslintrc.json') {
        expect(existsSync(join(appRoot, '.eslintrc.json'))).toBe(false)
      }

      filesToCheck.forEach((file) => {
        const fileContent = readFileSync(file).toString()
        expect(fileContent).toMatchSnapshot()
      })
    })
  }

  test('with prettier', async () => {
    setupTestProject('rn-setup-ci-yarn-flat')
    const { appRoot } = TEST_PROJECTS['rn-setup-ci-yarn-flat']

    const output = await cli([PRESET_FLAG, `--${FLAG}`, `--${PRETTIER_FLAG}`], {
      cwd: appRoot,
    })

    expect(output).toContain('Created ESLint workflow.')

    const packageJson = await getPackageJsonWithoutVersions(
      join(appRoot, 'package.json')
    )
    expect(packageJson).toMatchSnapshot()

    const eslintrcJson = readFileSync(
      join(appRoot, '.eslintrc.json')
    ).toString()
    expect(eslintrcJson).toMatchSnapshot()
  })
})
