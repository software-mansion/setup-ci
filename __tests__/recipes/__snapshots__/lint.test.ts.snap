// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint recipe rn-setup-ci-create-expo-stack lint 1`] = `
Object {
  "dependencies": Object {
    "react": "",
    "react-native": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@babel/preset-env": "",
    "@babel/runtime": "",
    "@react-native/babel-preset": "",
    "@react-native/eslint-config": "",
    "@react-native/metro-config": "",
    "@react-native/typescript-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "engines": Object {
    "node": ">=18",
  },
  "name": "RnSetupCiYarnFlat",
  "packageManager": "yarn@3.6.4",
  "private": true,
  "scripts": Object {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "start": "react-native start",
  },
  "version": "0.0.1",
}
`;

exports[`lint recipe rn-setup-ci-create-expo-stack lint 2`] = `
"name: Check ESLint

on:
  pull_request:

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🌿 Setup Node
        uses: actions/setup-node@v4
        with:
          node-version-file: '.nvmrc'
          cache: 'npm'

      - name: 📦 Install dependencies
        run: npm ci

      - name: 🤓 Run ESLint
        run: npm run lint
"
`;

exports[`lint recipe rn-setup-ci-create-expo-stack-bun lint 1`] = `
Object {
  "dependencies": Object {
    "react": "",
    "react-native": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@babel/preset-env": "",
    "@babel/runtime": "",
    "@react-native/babel-preset": "",
    "@react-native/eslint-config": "",
    "@react-native/metro-config": "",
    "@react-native/typescript-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "engines": Object {
    "node": ">=18",
  },
  "name": "RnSetupCiYarnFlat",
  "packageManager": "yarn@3.6.4",
  "private": true,
  "scripts": Object {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "start": "react-native start",
  },
  "version": "0.0.1",
}
`;

exports[`lint recipe rn-setup-ci-create-expo-stack-bun lint 2`] = `
"name: Check ESLint

on:
  pull_request:

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🥟 Setup Bun
        uses: oven-sh/setup-bun@v2
        with:
          bun-version-file: '.bun-version'

      - name: 📦 Install dependencies
        run: bun install --frozen-lockfile

      - name: 🤓 Run ESLint
        run: bun run lint
"
`;

exports[`lint recipe rn-setup-ci-create-expo-stack-pnpm lint 1`] = `
Object {
  "dependencies": Object {
    "react": "",
    "react-native": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@babel/preset-env": "",
    "@babel/runtime": "",
    "@react-native/babel-preset": "",
    "@react-native/eslint-config": "",
    "@react-native/metro-config": "",
    "@react-native/typescript-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "engines": Object {
    "node": ">=18",
  },
  "name": "RnSetupCiYarnFlat",
  "packageManager": "yarn@3.6.4",
  "private": true,
  "scripts": Object {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "start": "react-native start",
  },
  "version": "0.0.1",
}
`;

exports[`lint recipe rn-setup-ci-create-expo-stack-pnpm lint 2`] = `
"name: Check ESLint

on:
  pull_request:

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🐿️ Setup pnpm
        uses: pnpm/action-setup@v4
        with:
          version: 9

      - name: 🌿 Setup Node
        uses: actions/setup-node@v4
        with:
          node-version-file: '.nvmrc'
          cache: 'pnpm'

      - name: 📦 Install dependencies
        run: pnpm install --frozen-lockfile

      - name: 🤓 Run ESLint
        run: pnpm run lint
"
`;

exports[`lint recipe rn-setup-ci-npm-flat lint 1`] = `
Object {
  "dependencies": Object {
    "react": "",
    "react-native": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@babel/preset-env": "",
    "@babel/runtime": "",
    "@react-native/babel-preset": "",
    "@react-native/eslint-config": "",
    "@react-native/metro-config": "",
    "@react-native/typescript-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "engines": Object {
    "node": ">=18",
  },
  "name": "RnSetupCiYarnFlat",
  "packageManager": "yarn@3.6.4",
  "private": true,
  "scripts": Object {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "start": "react-native start",
  },
  "version": "0.0.1",
}
`;

exports[`lint recipe rn-setup-ci-npm-flat lint 2`] = `
"name: Check ESLint

on:
  pull_request:

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🌿 Setup Node
        uses: actions/setup-node@v4
        with:
          node-version-file: 'package.json'
          cache: 'npm'

      - name: 📦 Install dependencies
        run: npm ci

      - name: 🤓 Run ESLint
        run: npm run lint
"
`;

exports[`lint recipe rn-setup-ci-npm-flat lint 3`] = `
"{
  \\"parserOptions\\": {
    \\"ecmaVersion\\": \\"2020\\",
    \\"sourceType\\": \\"module\\"
  },
  \\"extends\\": [\\"@react-native\\", \\"eslint:recommended\\"],
  \\"env\\": {
    \\"browser\\": true,
    \\"node\\": true
  },
  \\"rules\\": {}
}
"
`;

exports[`lint recipe rn-setup-ci-npm-monorepo lint 1`] = `
Object {
  "dependencies": Object {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-font": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-splash-screen": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-updates": "",
    "expo-web-browser": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@react-native/eslint-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "jest-expo": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "main": "expo-router/entry",
  "name": "expo-app",
  "private": true,
  "scripts": Object {
    "android": "expo start --android",
    "ios": "expo start --ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "reset-project": "node ./scripts/reset-project.js",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`lint recipe rn-setup-ci-npm-monorepo lint 2`] = `
"name: Check ESLint

on:
  pull_request:
    paths:
      - apps/expo-app/**

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🌿 Setup Node
        uses: actions/setup-node@v4
        with:
          node-version-file: 'apps/expo-app/.nvmrc'
          cache: 'npm'

      - name: 📦 Install dependencies
        run: npm ci

      - name: 🤓 Run ESLint
        run: npm run lint
        working-directory: apps/expo-app
"
`;

exports[`lint recipe rn-setup-ci-npm-monorepo lint 3`] = `
"{
  \\"parserOptions\\": {
    \\"ecmaVersion\\": \\"2020\\",
    \\"sourceType\\": \\"module\\"
  },
  \\"extends\\": [\\"@react-native\\", \\"eslint:recommended\\"],
  \\"env\\": {
    \\"browser\\": true,
    \\"node\\": true
  },
  \\"rules\\": {}
}
"
`;

exports[`lint recipe rn-setup-ci-yarn-flat lint 1`] = `
Object {
  "dependencies": Object {
    "react": "",
    "react-native": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@babel/preset-env": "",
    "@babel/runtime": "",
    "@react-native/babel-preset": "",
    "@react-native/eslint-config": "",
    "@react-native/metro-config": "",
    "@react-native/typescript-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "engines": Object {
    "node": ">=18",
  },
  "name": "RnSetupCiYarnFlat",
  "packageManager": "yarn@3.6.4",
  "private": true,
  "scripts": Object {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "start": "react-native start",
  },
  "version": "0.0.1",
}
`;

exports[`lint recipe rn-setup-ci-yarn-flat lint 2`] = `
"name: Check ESLint

on:
  pull_request:

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🌿 Setup Node
        uses: actions/setup-node@v4
        with:
          node-version-file: 'package.json'
          cache: 'yarn'

      - name: 📦 Install dependencies
        run: yarn install --immutable

      - name: 🤓 Run ESLint
        run: yarn run lint
"
`;

exports[`lint recipe rn-setup-ci-yarn-flat lint 3`] = `
"{
  \\"parserOptions\\": {
    \\"ecmaVersion\\": \\"2020\\",
    \\"sourceType\\": \\"module\\"
  },
  \\"extends\\": [\\"@react-native\\", \\"eslint:recommended\\"],
  \\"env\\": {
    \\"browser\\": true,
    \\"node\\": true
  },
  \\"rules\\": {}
}
"
`;

exports[`lint recipe rn-setup-ci-yarn-monorepo lint 1`] = `
Object {
  "dependencies": Object {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-font": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-splash-screen": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-updates": "",
    "expo-web-browser": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@react-native/eslint-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "jest-expo": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "main": "expo-router/entry",
  "name": "expo-app",
  "private": true,
  "scripts": Object {
    "android": "expo start --android",
    "ios": "expo start --ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "reset-project": "node ./scripts/reset-project.js",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`lint recipe rn-setup-ci-yarn-monorepo lint 2`] = `
"name: Check ESLint

on:
  pull_request:
    paths:
      - apps/expo-app/**

jobs:
  eslint:
    name: ESLint
    runs-on: ubuntu-latest
    steps:
      - name: 🏗 Setup repo
        uses: actions/checkout@v4

      - name: 🌿 Setup Node
        uses: actions/setup-node@v4
        with:
          node-version-file: 'apps/expo-app/.nvmrc'
          cache: 'yarn'

      - name: 📦 Install dependencies
        run: yarn install --immutable

      - name: 🤓 Run ESLint
        run: yarn run lint
        working-directory: apps/expo-app
"
`;

exports[`lint recipe rn-setup-ci-yarn-monorepo lint 3`] = `
"{
  \\"parserOptions\\": {
    \\"ecmaVersion\\": \\"2020\\",
    \\"sourceType\\": \\"module\\"
  },
  \\"extends\\": [\\"@react-native\\", \\"eslint:recommended\\"],
  \\"env\\": {
    \\"browser\\": true,
    \\"node\\": true
  },
  \\"rules\\": {}
}
"
`;

exports[`lint recipe with prettier 1`] = `
Object {
  "dependencies": Object {
    "react": "",
    "react-native": "",
  },
  "devDependencies": Object {
    "@babel/core": "",
    "@babel/preset-env": "",
    "@babel/runtime": "",
    "@react-native/babel-preset": "",
    "@react-native/eslint-config": "",
    "@react-native/metro-config": "",
    "@react-native/typescript-config": "",
    "@types/react": "",
    "@types/react-test-renderer": "",
    "eslint": "",
    "eslint-plugin-ft-flow": "",
    "react-test-renderer": "",
    "typescript": "",
  },
  "engines": Object {
    "node": ">=18",
  },
  "name": "RnSetupCiYarnFlat",
  "packageManager": "yarn@3.6.4",
  "private": true,
  "scripts": Object {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint '**/*.{js,jsx,ts,tsx}'",
    "start": "react-native start",
  },
  "version": "0.0.1",
}
`;

exports[`lint recipe with prettier 2`] = `
"{
  \\"parserOptions\\": {
    \\"ecmaVersion\\": \\"2020\\",
    \\"sourceType\\": \\"module\\"
  },
  \\"extends\\": [
    \\"@react-native\\",
    \\"eslint:recommended\\",
    \\"prettier\\",
    \\"plugin:prettier/recommended\\"
  ],
  \\"env\\": {
    \\"browser\\": true,
    \\"node\\": true
  },
  \\"rules\\": {}
}
"
`;
